### PR TITLE
Unify proposals api

### PR DESCRIPTION
**Pull request by dkent600**
Resolves: #439 
Resolves: #445
Resolves: #446

Needed for:  https://github.com/daostack/arc.js/issues/116

Changes:
- `organizationsData` renamed to `organizationProposals` in `VestingScheme` and `VoteInOrganizationScheme`
- removes unused stuff in `GlobalConstraintRegistrar.Organization`struct
- makes `GlobalConstraintRegistrar` proposals accessible and consistent with the other schemes (as `organizationProposals`) (this is what is most needed by Arc.js)

### CODE REVIEW

File: contracts/VotingMachines/GenesisProtocol.sol [hunk 1]
```diff
@@ -77,7 +77,7 @@ contract GenesisProtocol is IntVoteInterface,UniversalScheme {
     mapping(bytes32=>Parameters) public parameters;  // A mapping from hashes to parameters
     mapping(bytes32=>Proposal) public proposals; // Mapping from the ID of the proposal to the proposal itself.
 
-    uint constant public NUM_OF_CHOICES = 2;
+    uint constant public MAX_NUM_OF_CHOICES = 2;
```
Comment (orenyodfat) [inline contracts/VotingMachines/GenesisProtocol.sol]: GenesisProtocol  currently support exactly 2 num of choices ..so MAX is not appropriate here.

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol [hunk 2]
```diff
@@ -35,16 +35,11 @@ contract GlobalConstraintRegistrar is UniversalScheme {
...
     }
 
-    // Struct holding the data for each organization
-    struct Organization {
-        bytes32 voteRegisterParams; // The voting parameters for adding a GC.
-        IntVoteInterface intVote; // The voting machine in which the voting takes place.
-        mapping(bytes32=>GCProposal) proposals; // A mapping from the proposal ID to the proposal itself.
-        mapping(address=>bytes32) voteToRemoveParams; // A mapping that saves the parameters for removing each GC.
-    }
+    // GCProposal by avatar and proposalId
```
Comment (orenyodfat) [inline contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Not sure what we gain having this modification.

File: package.json [hunk 3]
```diff
@@ -15,8 +15,8 @@
...
-    "ganache": "cross-conf-env ganache-cli -l npm_package_config_gasLimit --account=\"0x0191ecbd1b150b8a3c27c27010ba51b45521689611e669109e034fd66ae69621,9999999999999999999999999999999999999999999\" --account=\"0x00f360233e89c65970a41d4a85990ec6669526b2230e867c352130151453180d,9999999999999999999999999999999999999999999\" --account=\"0x987a26abca7432016104ce2f24ce639340e25afe06ac69f68791399e7a5d1028,9999999999999999999999999999999999999999999\" --account=\"0x89af34b1b7347834048b99423dad174a14bf14540d720d72c16ae92e94b987cb,9999999999999999999999999999999999999999999\" --account=\"0xc867be647eb2bc51e4c0d61066859875cf3634fe949b6f5f85c69ab90e485b37,9999999999999999999999999999999999999999999\" --account=\"0xefabcc2377dee5e51b5a9e65a3854aec85fbbec3cb584d8ad4f9679869fb33c6,9999999999999999999999999999999999999999999\"",
+    "test": "cross-var run-with-ganache --ganache-cmd ganache-cli --gasLimit $npm_package_config_gasLimit 'npm run truffle compile && npm run truffle migrate && npm run truffle test'",
```
Comment (orenyodfat) [inline package.json]: whats wrong with cross-conf-env ?

File: package.json [hunk 4]
```diff
@@ -37,7 +37,7 @@
     "babel-preset-es2015": "^6.24.1",
     "babel-register": "^6.26.0",
     "bignumber.js": "^5.0.0",
-    "cross-conf-env": "^1.1.2",
+    "cross-var": "^1.1.0",
```
Comment (orenyodfat) [inline package.json]: why  ?

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 1 above)
Comment (dkent600) [reply contracts/VotingMachines/GenesisProtocol.sol]: Perhaps then we can add MIN_NUM_OF_CHOICES too?

I need a consistent voting machine interface to be able to validate in Arc.js that a user's voting choices are within a range allowable by the voting machine.  It needs to do this in a voting machine-agnostic way.

Thus I am also proposing that MAX_NUM_OF_CHOICES (and MIN_NUM_OF_CHOICES) be made part of `IntVoteMachine` (see #444).

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 2 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Removes unused code and adds the ability to access proposals. See #466

File: package.json (hunk 3 above)
Comment (dkent600) [reply package.json]: Doesn't work on Windows.

File: package.json (hunk 4 above)
Comment (dkent600) [reply package.json]: Doesn't work on Windows.

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 1 above)
Comment (orenyodfat) [reply contracts/VotingMachines/GenesisProtocol.sol]: how about in arc.j : if GP so num of choices  =2 ?

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 2 above)
Comment (orenyodfat) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: how about access proposal by organizationsData[_avatar].proposals[_proposalId] ?
Which code is unused ?

File: package.json (hunk 3 above)
Comment (orenyodfat) [reply package.json]: windows is not supported by arc.

File: package.json (hunk 4 above)
Comment (orenyodfat) [reply package.json]: windows is not supported by arc.

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 2 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: `Organization.voteRegisterParams` and `Organization.IntVoteInterface intVote`

Note this change also makes `voteToRemoveParams` available, which it currently isn't.

Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: > how about access proposal by organizationsData[_avatar].proposals[_proposalId] 

This is explained in #466.  The proposals mapping is not exposed, not accessible.

File: package.json (hunk 3 above)
Comment (dkent600) [reply package.json]: Is there any reason not to make this simple change that will make my life easier (not to mention anyone else who would like to run the tests under windows)?

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 1 above)
Comment (dkent600) [reply contracts/VotingMachines/GenesisProtocol.sol]: > how about in arc.js : if GP so num of choices =2 ?

That doesn't work in a voting machine-agnostic context.  Refer to my comment just above.

File: package.json (hunk 3 above)
Comment (orenyodfat) [reply package.json]: It will make the project much complex to be verified and tested on different envs .

Comment (dkent600) [reply package.json]: I'm am not suggesting, by making this code change, that we change our policy on supporting Windows.  At the same time, I don't think our policy should be to actively resist improving the Windows experience if we are handed a trivial opportunity to do so.

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 2 above)
Comment (orenyodfat) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: please add tests for this change.

File: contracts/VotingMachines/GenesisProtocol.sol (hunk 1 above)
Comment (orenyodfat) [reply contracts/VotingMachines/GenesisProtocol.sol]: already addressed at #454

File: test/globalconstraintregistrar.js [hunk 5]
```diff
@@ -68,6 +68,19 @@ contract('GlobalConstraintRegistrar', function(accounts) {
...
+    it("proposeGlobalConstraint organizationsProposals", async function() {
+      var testSetup = await setup(accounts);
+      var globalConstraintMock = await GlobalConstraintMock.new();
+
+      var tx = await testSetup.globalConstraintRegistrar.proposeGlobalConstraint(testSetup.org.avatar.address,
+                                                                     globalConstraintMock.address,
+                                                                     "0x1234",
+                                                                     "0x1234");
+      var proposalId = await helpers.getValueFromLogs(tx, '_proposalId',1);
+      var organizationsProposals = await testSetup.globalConstraintRegistrar.organizationsProposals(testSetup.org.avatar.address,proposalId);
```
Comment (orenyodfat) [inline test/globalconstraintregistrar.js]: should be  var organizationProposal

Comment (dkent600) [reply test/globalconstraintregistrar.js]: Why is that since the contract property name is `organizationsProposals` ?

Comment (dkent600) [reply test/globalconstraintregistrar.js]: Do you want the contract mapping name to be everywhere changed to `organizationProposals`?

Comment (orenyodfat) [reply test/globalconstraintregistrar.js]: because this var hold a single proposal from a single avatar.

Comment (orenyodfat) [reply test/globalconstraintregistrar.js]: -the var in the test hold a single proposal from a single avatar -so it should be name accordingly. 
-the mapping in the contract is a map between organization**s**  to their proposal**s**  so it should be name accordingly.

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol [hunk 6]
```diff
@@ -35,16 +35,11 @@ contract GlobalConstraintRegistrar is UniversalScheme {
...
+    mapping(address=>mapping(bytes32=>GCProposal)) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/universalSchemes/GlobalConstraintRegistrar.sol]: organizationProposals -> organizationsProposals

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol [hunk 7]
```diff
@@ -35,16 +35,11 @@ contract GlobalConstraintRegistrar is UniversalScheme {
...
 
-    // A mapping from the organization (Avatar) address to the saved data of the organization:
-    mapping(address=>Organization) public organizationsData;
+    // voteToRemoveParams hash by avatar and proposal.gc
+    mapping(address=>mapping(address=>bytes32)) public voteToRemoveParams;
```
Comment (orenyodfat) [inline contracts/universalSchemes/GlobalConstraintRegistrar.sol]: why do we need additional mapping here?
voteToRemoveParams could be part of  organizationsProposals .

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 2 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Done

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 7 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Because those mappings are not accessible when contained inside the struct.  See #446.

Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Maybe you are suggesting moving the `voteToRemoveParams` _values_ (not the mapping) into the `GCProposal` struct?  That field actually is already there.  But the `voteToRemoveParams`  mapping is indexed by the GC address, unlike the `organizationProposals` mapping which is indexed by `proposalId`.  Such a mapping is needed by `proposeToRemoveGC`.

Comment (orenyodfat) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: How about just provide a getter function to make these mappings accessible ?

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 6 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Done

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 7 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: A function would not be consistent with the other schemes that create proposals that all have a mapping named `organizationsProposals`.

In any case, how would  a function be advantageous?

If the `voteToRemoveParams` mapping is not desired, then perhaps it doesn't need to be public.

Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: BTW, from the point of view of javascript, `organizationProposals` is already a function.

File: docs/generated_docs/universalSchemes/ContributionReward.md [hunk 8]
```diff
@@ -316,7 +316,7 @@ Returns:
 1. **output_0** *of type `address`*
 
 --- 
-### organizationsProposals(address,bytes32)
+### organizationProposals(address,bytes32)
```
[the remark this answers is not in the document]
Comment (dkent600) [reply docs/generated_docs/universalSchemes/ContributionReward.md]: I'm not real familiar with the Arc docs process.  What needs to be updated here?

[the remark this answers is not in the document]
Comment (orenyodfat) [reply docs/generated_docs/universalSchemes/ContributionReward.md]: so please remove this doc for the pr.

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 7 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: Done

File: contracts/universalSchemes/ContributionReward.sol [hunk 9]
```diff
@@ -45,7 +45,7 @@ contract ContributionReward is UniversalScheme {
     }
 
     // A mapping from the organization (Avatar) address to the saved data of the organization:
-    mapping(address=>mapping(bytes32=>ContributionProposal)) public organizationsProposals;
+    mapping(address=>mapping(bytes32=>ContributionProposal)) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/universalSchemes/ContributionReward.sol]: organizationProposals -> organizationsProposals 
please revert it back

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 6 above)
Comment (orenyodfat) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: It is still organizationProposals .
It should be organizationsProposals

File: contracts/universalSchemes/SchemeRegistrar.sol [hunk 10]
```diff
@@ -35,7 +35,7 @@ contract SchemeRegistrar is UniversalScheme {
     }
 
     // A mapping from the organization (Avatar) address to the saved data of the organization:
-    mapping(address=>mapping(bytes32=>SchemeProposal)) public organizationsProposals;
+    mapping(address=>mapping(bytes32=>SchemeProposal)) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/universalSchemes/SchemeRegistrar.sol]: please revert it back to organizationsProposals

File: contracts/universalSchemes/UpgradeScheme.sol [hunk 11]
```diff
@@ -34,7 +34,7 @@ contract UpgradeScheme is UniversalScheme, ExecutableInterface {
     }
 
     // A mapping from the organization's (Avatar) address to the saved data of the organization:
-    mapping(address=>mapping(bytes32=>UpgradeProposal)) public organizationsProposals;
+    mapping(address=>mapping(bytes32=>UpgradeProposal)) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/universalSchemes/UpgradeScheme.sol]: please revert it back to organizationsProposals

File: contracts/universalSchemes/VestingScheme.sol [hunk 12]
```diff
@@ -46,7 +46,7 @@ contract VestingScheme is UniversalScheme, ExecutableInterface {
     }
 
     // A mapping from the organization (Avatar) address to the saved data of the organization:
-    mapping(address=>mapping(bytes32=>Agreement)) public organizationsData;
+    mapping(address=>mapping(bytes32=>Agreement)) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/universalSchemes/VestingScheme.sol]: please revert it back to organizationsProposals

File: contracts/universalSchemes/VoteInOrganizationScheme.sol [hunk 13]
```diff
@@ -30,7 +30,7 @@ contract VoteInOrganizationScheme is UniversalScheme, ExecutableInterface, Actio
     }
 
     // A mapping from the organization (Avatar) address to the saved data of the organization:
-    mapping(address=>mapping(bytes32=>VoteProposal)) public organizationsData;
+    mapping(address=>mapping(bytes32=>VoteProposal)) public organizationProposals;
```
Comment (orenyodfat) [inline contracts/universalSchemes/VoteInOrganizationScheme.sol]: organizationProposals - > organizationsProposals

File: contracts/universalSchemes/ContributionReward.sol (hunk 9 above)
Comment (dkent600) [reply contracts/universalSchemes/ContributionReward.sol]: I'm not sure where this diff is coming from, it is name `organizationProposals` [in the code](https://github.com/daostack/arc/blob/eb7b6965105d32dec990fb6409f4841d86944ad7/contracts/universalSchemes/ContributionReward.sol#L48).

File: contracts/universalSchemes/SchemeRegistrar.sol (hunk 10 above)
Comment (dkent600) [reply contracts/universalSchemes/SchemeRegistrar.sol]: I'm not sure where this diff is coming from, it is named `organizationProposals` in the code.

File: contracts/universalSchemes/UpgradeScheme.sol (hunk 11 above)
Comment (dkent600) [reply contracts/universalSchemes/UpgradeScheme.sol]: I'm not sure where this diff is coming from, it is named `organizationProposals` in the code.

File: contracts/universalSchemes/VestingScheme.sol (hunk 12 above)
Comment (dkent600) [reply contracts/universalSchemes/VestingScheme.sol]: I'm not sure where this diff is coming from, it is named `organizationProposals` in the code.

File: contracts/universalSchemes/VoteInOrganizationScheme.sol (hunk 13 above)
Comment (dkent600) [reply contracts/universalSchemes/VoteInOrganizationScheme.sol]: I'm not sure where this diff is coming from, it is named `organizationProposals` in the code.

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 6 above)
Comment (dkent600) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: It looks OK to me:  https://github.com/daostack/arc/tree/unifyProposalsAPI/contracts/universalSchemes

File: contracts/universalSchemes/ContributionReward.sol (hunk 9 above)
Comment (orenyodfat) [reply contracts/universalSchemes/ContributionReward.sol]: it should be name organizationsProposals

File: contracts/universalSchemes/GlobalConstraintRegistrar.sol (hunk 6 above)
Comment (orenyodfat) [reply contracts/universalSchemes/GlobalConstraintRegistrar.sol]: It should be name organizationsProposals

File: contracts/universalSchemes/SchemeRegistrar.sol (hunk 10 above)
Comment (orenyodfat) [reply contracts/universalSchemes/SchemeRegistrar.sol]: it should be name organizationsProposals

File: contracts/universalSchemes/UpgradeScheme.sol (hunk 11 above)
Comment (orenyodfat) [reply contracts/universalSchemes/UpgradeScheme.sol]: It should be name organizationsProposals

File: contracts/universalSchemes/VestingScheme.sol (hunk 12 above)
Comment (orenyodfat) [reply contracts/universalSchemes/VestingScheme.sol]: It should be name organizationsProposals

File: contracts/universalSchemes/VoteInOrganizationScheme.sol (hunk 13 above)
Comment (orenyodfat) [reply contracts/universalSchemes/VoteInOrganizationScheme.sol]: It should be name organizationsProposals

Comment (dkent600) [reply contracts/universalSchemes/VoteInOrganizationScheme.sol]: ok, reverted back, except for the test code variable names

File: test/voteinorganization.js [hunk 14]
```diff
@@ -108,9 +108,9 @@ contract('VoteInOrganizationScheme', function(accounts) {
                                                                  originalProposalId);
              var proposalId = await helpers.getValueFromLogs(tx, '_proposalId');
              await testSetup.voteInOrganizationParams.votingMachine.absoluteVote.vote(proposalId,0,{from:accounts[2]});
-             //check organizationsData after execution
-             var organizationsData = await testSetup.voteInOrganization.organizationsData(testSetup.org.avatar.address,proposalId);
-             assert.equal(organizationsData[0],0x0000000000000000000000000000000000000000);//new contract address
+             //check organizationsProposals after execution
+             var organizationProposals = await testSetup.voteInOrganization.organizationsProposals(testSetup.org.avatar.address,proposalId);
```
Comment (orenyodfat) [inline test/voteinorganization.js]: organizationProposals -> organizationProposal

File: test/upgradescheme.js [hunk 15]
```diff
@@ -128,15 +128,15 @@ contract('UpgradeScheme', function(accounts) {
              //Vote with reputation to trigger execution
              var proposalId = await helpers.getValueFromLogs(tx, '_proposalId',1);
              //check organizationsProposals before execution
-             var organizationsProposals = await testSetup.upgradeScheme.organizationsProposals(testSetup.org.avatar.address,proposalId);
-             assert.equal(organizationsProposals[0],newController.address);//new contract address
-             assert.equal(organizationsProposals[2].toNumber(),1);//proposalType
+             var organizationProposals = await testSetup.upgradeScheme.organizationsProposals(testSetup.org.avatar.address,proposalId);
```
Comment (orenyodfat) [inline test/upgradescheme.js]: organizationProposals - > organizationProposal

**Pull request merged**